File: mobile/lib/features/post/domain/entities/post.dart
class Post {
  final String id;
  final String authorId;
  final String title;
  final String slug;
  final String excerpt;
  final String content;
  final String contentMd;
  final String? coverImageUrl;
  final String? category;
  final List<String> tags;
  final String status;
  final bool isPremium;
  final String? language;
  final int? wordCount;
  final int? readingTime;
  final DateTime? publishedAt;
  final DateTime createdAt;
  final Author? author;
  final int commentCount;
  final Map<String, int> reactions;
  final List<String> viewerReactions;
  final bool isBookmarked;

  Post({
    required this.id,
    required this.authorId,
    required this.title,
    required this.slug,
    required this.excerpt,
    required this.content,
    this.contentMd = '',
    this.coverImageUrl,
    this.category,
    this.tags = const [],
    required this.status,
    this.isPremium = false,
    this.language,
    this.wordCount,
    this.readingTime,
    this.publishedAt,
    required this.createdAt,
    this.author,
    this.commentCount = 0,
    this.reactions = const {},
    this.viewerReactions = const [],
    this.isBookmarked = false,
  });

  int get likeCount {
    var total = 0;
    for (final count in reactions.values) {
      total += count;
    }
    return total;
  }

  bool get viewerLiked =>
      viewerReactions.contains('❤️') || viewerReactions.contains('like');

  factory Post.fromJson(Map<String, dynamic> json) {
    return Post(
      id: json['id'] as String,
      authorId: json['author_id'] as String,
      title: json['title'] as String,
      slug: json['slug'] as String,
      excerpt: json['excerpt'] as String? ?? '',
      content: json['content'] as String? ?? '',
      contentMd: json['content_md'] as String? ?? '',
      coverImageUrl: json['cover_image_url'] as String?,
      category: json['category'] as String?,
      tags: (json['tags'] as List<dynamic>?)?.map((e) => e as String).toList() ?? [],
      status: json['status'] as String? ?? 'draft',
      isPremium: json['is_premium'] as bool? ?? false,
      language: json['language'] as String?,
      wordCount: (json['word_count'] as num?)?.toInt(),
      readingTime: (json['reading_time'] as num?)?.toInt(),
      publishedAt: json['published_at'] != null ? DateTime.parse(json['published_at'] as String) : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      author: json['author'] != null ? Author.fromJson(json['author'] as Map<String, dynamic>) : null,
      commentCount: (json['comment_count'] as num?)?.toInt() ?? 0,
      reactions: _parseReactions(json['reactions']),
      viewerReactions: (json['viewer_reactions'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          const [],
      isBookmarked: json['is_bookmarked'] as bool? ?? false,
    );
  }

  static Map<String, int> _parseReactions(dynamic raw) {
    if (raw is! Map) return {};
    return raw.map((k, v) => MapEntry(k.toString(), (v as num).toInt()));
  }
}

class Author {
  final String id;
  final String username;
  final String displayName;
  final String? avatarUrl;
  final String? bio;
  final bool isVerified;

  Author({
    required this.id,
    required this.username,
    required this.displayName,
    this.avatarUrl,
    this.bio,
    this.isVerified = false,
  });

  factory Author.fromJson(Map<String, dynamic> json) {
    return Author(
      id: json['id'] as String,
      username: json['username'] as String,
      displayName: json['display_name'] as String,
      avatarUrl: json['avatar_url'] as String?,
      bio: json['bio'] as String?,
      isVerified: json['is_verified'] as bool? ?? false,
    );
  }
}

File: backend/internal/post/model/post.go
package model

import (
	"errors"
	"time"

	"github.com/lib/pq"
	authmodel "github.com/xilo-platform/xilo/internal/auth/model"
)

type Post struct {
	ID           string     `json:"id" db:"id"`
	AuthorID     string     `json:"author_id" db:"author_id"`
	Title        string     `json:"title" db:"title"`
	Slug         string     `json:"slug" db:"slug"`
	Excerpt      string     `json:"excerpt" db:"excerpt"`
	Content      string     `json:"content" db:"content"`
	ContentMD    string     `json:"content_md" db:"content_md"`
	CoverImageURL *string   `json:"cover_image_url,omitempty" db:"cover_image_url"`
	Category     *string    `json:"category,omitempty" db:"category"`
	Tags         pq.StringArray `json:"tags" db:"tags"`
	Status       string     `json:"status" db:"status"`
	IsPremium    bool       `json:"is_premium" db:"is_premium"`
	WordCount    int        `json:"word_count" db:"word_count"`
	ReadingTime  int        `json:"reading_time" db:"reading_time"`
	Language     string     `json:"language" db:"language"`
	ViewCount    int64      `json:"view_count" db:"view_count"`
	ScheduledAt  *time.Time `json:"scheduled_at,omitempty" db:"scheduled_at"`
	PublishedAt  *time.Time `json:"published_at,omitempty" db:"published_at"`
	CreatedAt    time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt    time.Time  `json:"updated_at" db:"updated_at"`
	DeletedAt    *time.Time `json:"deleted_at,omitempty" db:"deleted_at"`

	Author           *authmodel.User   `json:"author,omitempty" db:"-"`
	CommentCount     int               `json:"comment_count" db:"-"`
	RepostCount      int               `json:"repost_count" db:"-"`
	Reactions        map[string]int    `json:"reactions,omitempty" db:"-"`
	ViewerReactions  []string          `json:"viewer_reactions,omitempty" db:"-"`
	IsBookmarked     bool              `json:"is_bookmarked" db:"-"`
	IsReposted       bool              `json:"is_reposted" db:"-"`
}

var ErrInvalidViewSession = errors.New("session_id is required for anonymous views")

type RecordViewRequest struct {
	SessionID string `json:"session_id"`
}

type RecordViewResult struct {
	Counted   bool   `json:"counted"`
	ViewCount int64  `json:"view_count"`
	Slug      string `json:"-"`
	AuthorID  string `json:"-"`
}

type PostVersion struct {
	ID        string    `json:"id" db:"id"`
	PostID    string    `json:"post_id" db:"post_id"`
	Title     string    `json:"title" db:"title"`
	Content   string    `json:"content" db:"content"`
	ContentMD string    `json:"content_md" db:"content_md"`
	Version   int       `json:"version" db:"version"`
	CreatedAt time.Time `json:"created_at" db:"created_at"`
}

type CreatePostRequest struct {
	Title        string   `json:"title"`
	Slug         string   `json:"slug"`
	Excerpt      string   `json:"excerpt"`
	Content      string   `json:"content"`
	ContentMD    string   `json:"content_md"`
	CoverImageURL string  `json:"cover_image_url"`
	Category     string   `json:"category"`
	Tags         []string `json:"tags"`
	Status       string   `json:"status"`
	IsPremium    bool     `json:"is_premium"`
	Language     string   `json:"language"`
	ScheduledAt  *time.Time `json:"scheduled_at"`
}

type UpdatePostRequest struct {
	Title        *string   `json:"title"`
	Slug         *string   `json:"slug"`
	Excerpt      *string   `json:"excerpt"`
	Content      *string   `json:"content"`
	ContentMD    *string   `json:"content_md"`
	CoverImageURL *string  `json:"cover_image_url"`
	Category     *string   `json:"category"`
	Tags         *[]string `json:"tags"`
	Status       *string   `json:"status"`
	IsPremium    *bool     `json:"is_premium"`
	Language     *string   `json:"language"`
	ScheduledAt  *time.Time `json:"scheduled_at"`
}

type PostListParams struct {
	Cursor   string
	Limit    int
	Category string
	Tag      string
	Author    string
	Status    string
	Language  string
	MediaOnly bool
	ViewerID  string
}

// TagSuggestion is a hashtag/tag used for autocomplete and trending.
type TagSuggestion struct {
	Tag   string `json:"tag" db:"tag"`
	Count int64  `json:"count" db:"count"`
}
